ariadne/cardano/test/backend/Test/Spec/AcidState.hs

```diff
@@ -0,0 +1,103 @@
+module Test.Spec.AcidState (spec) where
+
+import Control.Exception.Base (ErrorCall(..), handle)
+import Control.Monad.Component hiding (throwM)
+
+import System.IO.Temp (withSystemTempDirectory)
+
+import Data.Acid (closeAcidState, openLocalStateFrom, query)
+
+import Data.Text.Buildable (build)
+import Data.Text.Lazy.Builder (toLazyText)
+
+import Test.Hspec (Spec, describe)
+import Test.Hspec.QuickCheck (prop)
+import Test.QuickCheck (arbitrary, withMaxSuccess)
+import Test.QuickCheck.Monadic (monadicIO, pick)
+
+import Test.Spec.CreateWallet (NewWallet, applyNewWallet, genNewWalletRq)
+import Test.Spec.Fixture (genSpendingPassword, withLayerLocalStorage)
+
+import Ariadne.Wallet.Cardano.Kernel.DB.AcidState (DB(..), Snapshot(..), defDB)
+import Ariadne.Wallet.Cardano.Kernel.DB.HdWallet (HdWallets(..))
+import Ariadne.Wallet.Cardano.Kernel.DB.Util.IxSet (IxSet, size)
+import Ariadne.Wallet.Cardano.Kernel.Internal (PassiveWallet(..), wallets)
+import Ariadne.Wallet.Cardano.WalletLayer (PassiveWalletLayer, pwlCreateWallet)
+
+spec :: Spec
+spec = describe "AcidState" $ do
+  prop "can open empty database" $ withMaxSuccess 1 $
+    monadicIO $ do
+      pm <- pick arbitrary
+      liftIO $ withSystemTempDirectory "testWalletDBEmpty" $ \path -> do
+        handle emptyDBErrorCatch $

```

Comment:
Do we need it?

Reply:
I think yes.  
Otherwise we get only an exception from the second property (i think that exception is not a valid result of test). 
We can catch it in the second propety, but in that case we would not know why the test failed - because Ariadne can't open empty database or it loads acid db state in a wrong way.

Reply:
I thought about it yesterday and think that we should leave both peroperties

Reply:
I don't get it.
If `withLayerLocalStorage @IO pm path $ \_ wallet -> checkEmptyDB wallet` throws `ComponentError`, it will be caught and the result will be `False` which is the same as if `checkEmptyDB` returns `False`.
So we can't distinguish whether the DB couldn't be open or it was open with unexpected state, can we?

Reply:
OTOH if we don't catch anything here, in the first case we'll get an exception (which is good, because the test should fail anyway) with error message and in the second case we'll get `Falsifiable` which would mean that `checkEmptyDB` returned `False`.

Reply:
> Oh. I thought that you say about the whole first property.

No, I mean exactly this line which handles a particular exception.

I am going to approve this PR because it already took a lot of time and the bug is critical and the things we are discussing here are quite minor. However, it might make sense to open a follow-up PR.

Reply:
> I think that exception is a bad way to say that a test failed. I think that a test should return Testable that inform a programmer that a property is not satisfied anymore. An exception is a way to said that something went wrong during a test, and it is not valid or could not be done.

My opinion on that in a nutshell:
1. I don't see anything bad in having a test fail by means of exception. When you write a test and don't catch a certain exception it means that you expect that this exception won't be thrown. Testing framework takes care of handling exceptions that your code doesn't catch.
2. As I said, in this particular case the only change visible to the person running this test is that they no longer can distinguish failure due to wrong contents of DB and failure to open DB. Which is quite bad.
3. You could avoid (2) if instead of returning `False` you were using `fail` inside `PropertyM` monad. But the end result is almost the same as not catching the exception at all. The only benefit of doing so is that you could modify the error message. And the disadvantage is that it makes the test significantly larger and more complex.
4. We don't expect this exception to be thrown anymore. So it's essentially the same as any other exception. Any other exception can be thrown, but we don't catch them. We could catch all exceptions, but fortunately we don't have to, because it's done for us by the testing engine.

> I think now, that I should make 3 properties - opening DB, an emptiness of an empty DB and the equality of the closed and opened state.

I think it would be a move into wrong direction. First two tests would be doing basically the same thing: open empty DB. Instead of doing the same thing twice, results of the test can be made more specific. As I said, if you are strictly opposed to handling it by means of builtin runtime exceptions, you can do it using `fail` inside the `PropertyM` monad.